### PR TITLE
feat: prepare jsr publish

### LINKED ISSUE #980: feat: prepare for jsr.io publishing
[jsr.io](https://jsr.io/) is a special registry for TS packages publishing. Lets add a compatibility test and make it green.

```js
  "scripts": {
    // ...
    "test:jsr": "jsr publish --dry-run",
  },
```

See: 
* https://github.com/webpod/zurk/blob/main/src/scripts/build-jsr.mjs
* https://github.com/webpod/zurk/blob/main/package.json#L59

### PR DESCRIPTION
**Pull request by easymikey**
Fixes #980 

### Description

To support JSR:
- [x] create build jsr script
- [x] add tests
- [x] add return types
- [x] add comment for fix publish in future 

Screenshot 2024-12-15 at 17 36 36

- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by antongolub** [comment]
Hey, @easymikey,

1) Let's keep the `core.ts` formatting
2) CLA is required step

File: src/core.ts [hunk 1]
```diff
@@ -284,43 +286,43 @@ export class ProcessPromise extends Promise<ProcessOutput> {
...
-              status,
-              signal,
-              dto.stderr(),
-              self._from
-            )),
-            ...error && {
-              code: () => null,
-              signal: () => null,
-              message: () => ProcessOutput.getErrorMessage(error, self._from)
+            self._resolved = true
```
Comment (antongolub) [inline src/core.ts:289]: Let's keep the formatting for valid git blame.

**Review by antongolub — CHANGES_REQUESTED** [review]

File: src/core.ts (hunk 1 above)
Comment (easymikey) [reply src/core.ts:272]: Done

**Comment by easymikey** [comment]
@antongolub what do you think about adding test:coverege job to pull request action?

**Comment by antongolub** [comment]
> @antongolub what do you think about adding test:coverege job to pull request action?

We've already done.  https://github.com/google/zx/blob/14ff473c3aae2d8116c1fde8d94c369ae550093c/.github/workflows/test.yml#L89

File: src/core.ts (hunk 1 above)
Comment (antongolub) [reply src/core.ts:272]: **core.ts**
image

Comment (easymikey) [reply src/core.ts]: Sorry, I rebase another commit, everything is ok now

**Comment by easymikey** [comment]
> > @antongolub what do you think about adding test:coverege job to pull request action?
> 
> We've already done.
> 
> https://github.com/google/zx/blob/14ff473c3aae2d8116c1fde8d94c369ae550093c/.github/workflows/test.yml#L89

Tell me, why are they launched only by the maintainer?

File: src/cli.ts [hunk 2]
```diff
@@ -57,19 +57,18 @@ export function printUsage() {
...
    --cwd=<path>         set current directory
-   --eval=<js>, -e      evaluate script 
+   --eval=<js>, -e      evaluate script
    --ext=<.mjs>         default extension
    --install, -i        install dependencies
    --version, -v        print current zx version
    --help, -h           print help
    --repl               start repl
    --experimental       enables experimental features (deprecated)
-
```
Comment (antongolub) [inline src/cli.ts:67]: Let's keep this extra line.

File: scripts/build-jsr.mjs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import fs from 'node:fs'
+import path from 'node:path'
+const cwd = process.cwd()
```
Comment (antongolub) [inline scripts/build-jsr.mjs:17]: Let's use root-based resolve, like here https://github.com/google/zx/blob/9f15fd4810738f7f7c838465c52d3add3844c25d/scripts/build-tests.mjs#L22

**Review by antongolub — CHANGES_REQUESTED** [review]
Almost done. Just a few suggestions

File: scripts/build-jsr.mjs (hunk 3 above)
Comment (easymikey) [reply scripts/build-jsr.mjs]: d1

File: src/cli.ts (hunk 2 above)
Comment (easymikey) [reply src/cli.ts:67]: d1

File: scripts/build-jsr.mjs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import fs from 'node:fs'
+import path from 'node:path'
+const __dirname = path.dirname(new URL(import.meta.url).pathname)
+const root = path.resolve(__dirname, '..')
+const pkgJson = JSON.parse(
+  fs.readFileSync(path.resolve(root, 'package.json'), 'utf-8')
+)
+
+fs.writeFileSync(
+  path.resolve(cwd, 'jsr.json'),
```
Comment (antongolub) [inline scripts/build-jsr.mjs:24]: cwd > root

File: package.json [hunk 5]
```diff
@@ -70,6 +70,7 @@
     "pretest": "npm run build",
     "test": "npm run test:size && npm run fmt:check && npm run test:unit && npm run test:types && npm run test:license",
```
Comment (antongolub) [inline package.json:71]: 1) "test": "... && npm run test:jsr"
2) add `jsr.json` to `.gitignore`

Comment (antongolub) [reply package.json:71]: Ahhh... its in temp. Ok, nvm

File: scripts/build-jsr.mjs (hunk 4 above)
Comment (easymikey) [reply scripts/build-jsr.mjs:24]: d1

**Review by antongolub — APPROVED** [review]
lgtm. Thanks for the contribution!

**Comment by easymikey** [comment]
@easymikey Thanks for your patience :)

**Comment by easymikey** [comment]
@antongolub  Thanks for your patience :)

**Pull request merged**